src/librustc_session/utils.rs

```diff
@@ -12,17 +12,19 @@ impl Session {
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, RustcEncodable, RustcDecodable)]
 pub enum NativeLibKind {
-    /// Static library (`libfoo.a` archive) included when linking a final binary,
+    /// Static library (`libfoo.a` or `foo.lib` on Windows) included when linking a final binary,
     /// but not when archiving an rlib.
     StaticNoBundle,
-    /// Static library (`libfoo.a` archive) included when linking a final binary,
+    /// Static library (`libfoo.a` or `foo.lib` on Windows) included when linking a final binary,
     /// but also included when archiving an rlib.
     StaticBundle,
+    /// Dynamic library (`libfoo.so`) or an import library (`foo.lib`) on Windows.
+    Dylib,
     /// Windows dynamic library (`foo.dll`) without a corresponding import library (`foo.lib`).
     RawDylib,
     /// A macOS-specific kind of dynamic libraries.
     Framework,
-    /// The library kind wasn't specified, dynamic linking is currently preferred.
+    /// The library kind wasn't specified, `Dylib` is currently used as a default.
```

Comment:
Hm I'm not sure I quite follow this comment -- when do we encounter Unspecified if dylib is the default? Or does this mean that (as I think) that currently when we encounter an unspecified library kind, we store it as unspecified, but the code treats it equivalently to Dylib?

If that's right, it seems like we might want to adjust this such that `Dylib` is instead something like `Dylib { explicit: bool }` perhaps? I'm not sure whether that's worthwhile though given how rarely we add things, but I'd be a bit concerned that over time we may not be actually treating unspecified entirely equivalently to unspecified.

Reply:
`Unspecified` and `Dylib` behave identically right now, but they may behave differently if new options are added (https://github.com/rust-lang/rust/issues/39998) or experiments performed (https://github.com/rust-lang/rust/issues/71647).
So the intent is indeed to add things and this PR is a preparation.